prow/config/secrets_agent.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// SecretAgent watches a path and automatically loads the config stored
+// therein.
+type SecretAgent struct {
+	sync.Mutex
+	secretsMap map[string][]byte
+}
+
+func NewSecretAgent() *SecretAgent {
+	return &SecretAgent{
+		secretsMap: make(map[string][]byte),
+	}
+}
+
+// Start will begin polling the config file at the path. If the first load
+// fails, Start with return the error and abort. Future load failures will log
+// the failure message but continue attempting to load.
+func (sa *SecretAgent) Start(paths []string) error {
+	secretsMap, err := LoadSecrets(paths)
+	if err != nil {
+		return err
+	}
+
+	sa.secretsMap = secretsMap
+
+	go func() {
+		var lastModTime time.Time
+
+		skips := 0
+		for range time.Tick(1 * time.Second) {
+			for secretPath := range secretsMap {
+				if skips < 600 {
+					// Check if the file changed to see if it needs to be re-read.
+					secretStat, err := os.Stat(secretPath)
+					if err != nil {
+						logrus.WithField("secretPath", secretPath).WithError(err).Error("Error loading secret file.")
+						continue
+					}
+
+					recentModTime := secretStat.ModTime()
+
+					if !recentModTime.After(lastModTime) {
+						skips++
+						continue // file hasn't been modified
+					}
+					lastModTime = recentModTime
+				}
+
+				if secretsMap, err := LoadSecrets(paths); err != nil {
+					logrus.WithField("Secret paths", paths).
+						WithError(err).Error("Error loading config.")
+				} else {
+					skips = 0
+					sa.Lock()
+					sa.secretsMap = secretsMap
+					sa.Unlock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Config returns the latest config. Do not modify the config.
+func (sa *SecretAgent) Config() map[string][]byte {
+	sa.Lock()
+	defer sa.Unlock()
+	return sa.secretsMap
+}
+
+// Set sets the config. Useful for testing.
+func (sa *SecretAgent) Set(secretsMap map[string][]byte) {
```

Comment:
If this is only for testing, can we use an interface for the SecretAgent and have this method only exit on the test mock? There is no case for production code to set the secret data, right?

Reply:
After the secret will change, the agent will set the new data. I didn't use Set() there, but I changed it after your suggestion.